### ISSUE TITLE
False-negative Style/ConditionalAssignment cop

### ISSUE THREAD

**Issue by taichi-ishitani** [rubocop/rubocop#12216]
There is a case that `Style/ConditionalAssignment` cop reports a false-negative warning.

Here is an example:

```ruby
class Foo
  attr_reader :v

  def ==(other)
    case other
    when self.class then v == other.v
    else v == other
    end
  end
end
```

However, the cop reprots no warnings for following exmaples.

```ruby
class Bar
  attr_reader :v

  def ==(other)
    case other
    when self.class then other.v == v
    else v == other
    end
  end
end

class Baz
  attr_reader :v

  def ==(other)
    case other
    when self.class then v == other.v
    end
  end
end
```

--------

## Expected behavior

The cop should not report warnings.

## Actual behavior

The cop reports a warning.

```
$ rubocop --debug test.rb
For /home/taichi/temp: Default configuration from /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.56.3/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /home/taichi/temp/test.rb
Loading cache from /home/taichi/.cache/rubocop_cache/003ba08eb40ef09d8bd4a8d424addcca9e8d90b4/6d7a3b621ca1730e04accd938619e4bdab66cfb1/d52ad35602ca393ad38120bd1f3d627eedbb6a8d
C

Offenses:

test.rb:8:5: C: [Correctable] Style/ConditionalAssignment: Use the return of the conditional for variable assignment and comparison.
    case other ...
    ^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense autocorrectable
Finished in 0.08229890300026454 seconds
```

## Steps to reproduce the problem

Run rubocop with the following example code.
Then you will see the cop reports a warning for the `Foo#==` method but not the `Bar#==` and `Baz#==` methods.

```ruby
# frozen_string_literal: true

# Foo class
class Foo
  attr_reader :v

  def ==(other)
    case other
    when self.class then v == other.v
    else v == other
    end
  end
end

# Bar class
class Bar
  attr_reader :v

  def ==(other)
    case other
    when self.class then other.v == v
    else v == other
    end
  end
end

# Baz class
class Baz
  attr_reader :v

  def ==(other)
    case other
    when self.class then v == other.v
    end
  end
end
```

## RuboCop version

```
$ rubocop -V
1.56.3 (using Parser 3.2.2.3, rubocop-ast 1.29.0, running on ruby 3.2.2) [x86_64-linux]
```

**Comment by fatkodima** [comment]
Expected behavior to me. 

1st example: rubocop wants you to write it like this:
```ruby
class Foo
  attr_reader :v

  def ==(other)
    v == case other
         when self.class then other.v
         else other
         end
  end
end
```

2nd example: we have `v` in one case at a right hand side, in another - as left hand side. Since `==` is a method call, `a == b` can be not the same as `b == a`, so that's why it is not considered. Or just to make the code a little smaller.

3rd example: there is no `else` branch and this is not convertible.

**Comment by taichi-ishitani** [comment]
Hi @fatkodima ,
Thank you for your comment!
I understood why rubocop reports a warning for my example code.
